streams/src/main/java/org/apache/kafka/streams/state/VersionedKeyValueStore.java

```diff
@@ -0,0 +1,99 @@
+ * {@link #put(Object, Object, long)} with the same key and timestamp will replace the first.
+ * <p>
+ * Each store instance has an associated, fixed-duration "history retention" which specifies
+ * how long old record versions should be kept for. In particular, a versioned store guarantees
+ * to return accurate results for calls to {@link #get(Object, long)} where the provided timestamp
+ * bound is within history retention of the current observed stream time. (Queries with timestamp
+ * bound older than the specified history retention are considered invalid.)
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public interface VersionedKeyValueStore<K, V> extends StateStore {
+
+    /**
+     * Add a new record version associated with this key.
+     *
+     * @param key       The key
+     * @param value     The value, it can be {@code null};
+     *                  if the serialized bytes are also {@code null} it is interpreted as a delete
+     * @param timestamp The timestamp for this record version
+     * @throws NullPointerException If {@code null} is used for key.
+     */
+    void put(K key, V value, long timestamp);
+
+    /**
+     * Delete the value associated with this key from the store, at the specified timestamp
+     * (if there is such a value), and return the deleted value.
+     * <p>
+     * This operation is semantically equivalent to {@link #get(Object, long)} #get(key, timestamp))}
+     * followed by {@link #put(Object, Object, long) #put(key, null, timestamp)}.
+     *
+     * @param key       The key
+     * @param timestamp The timestamp for this delete
+     * @return The value and timestamp of the latest record associated with this key
+     *         as of the deletion timestamp (inclusive), or {@code null} if any of
+     *         (1) the store contains no records for this key, (2) the latest record
+     *         for this key as of the deletion timestamp is a tombstone, or
+     *         (3) the deletion timestamp is older than this store's history retention
+     *         (i.e., this store no longer contains data for the provided timestamp).
+     * @throws NullPointerException If {@code null} is used for key.
+     */
+    VersionedRecord<V> delete(K key, long timestamp);
+
+    /**
+     * Get the latest (by timestamp) record associated with this key.
+     *
+     * @param key The key to fetch
+     * @return The value and timestamp of the latest record associated with this key, or
+     *         {@code null} if either (1) the store contains no records for this key or (2) the
+     *         latest record for this key is a tombstone.
+     * @throws NullPointerException       If null is used for key.
+     * @throws InvalidStateStoreException if the store is not initialized
+     */
+    VersionedRecord<V> get(K key);
+
+    /**
+     * Get the latest record associated with this key with timestamp not exceeding the specified
+     * timestamp bound.
```

Comment:
Hm. Currently "validFrom" and "validTo" are not mentioned anywhere in the public-facing interfaces (VersionedKeyValueStore, VersionedRecord, etc) in javadocs or as parameter names; they are only "officially" introduced as concepts in the javadocs for RocksDBVersionedStore and its implementation helpers. I'm a bit hesitant to introduce these concepts in the public-facing docs at this time if the only purpose is to clarify what's returned from `get(key, asOfTimestamp)`, particularly because during KIP discussion one person mentioned that these terms did not feel intuitive to them. If you feel strongly though, I can move the explanation of "validity interval" including "validFrom" and "validTo" from the javadocs for RocksDBVersionedStore to VersionedKeyValueStore itself, and then update this javadoc.

Reply:
Fair point.

> particularly because during KIP discussion one person mentioned that these terms did not feel intuitive to them

I cannot remember this; personally I always found it very intuitive (also because it's use in temporal tables in SQL), but of course this could just be me.

I agree that we should not use it only here, but rather introduce it "globally" or not use it at all. It's "just" JavaDocs and just easy to change in follow up releases if thy cause confusion, so I don't feel strong about it and leave it to your judgement.

Reply:
Makes sense. I'll omit for now, and we can always revisit in the future.